### DISCUSSION TITLE
How would I go about making a server-side mod that prevents the use of health/attribute mods like Mob Plaques on the client?

### DISCUSSION THREAD

**Discussion by TechPro424** [FabricMC/fabric-api#2851] [Mod Dev Support]
How would I go about making a server-side mod that prevents the use of health/attribute mods like [Mob Plaques](https://modrinth.com/mod/mob-plaques) on the client?
I tried mixing into the head of the `getHealth()` method in the `HealthUpdateS2CPacket` class so that it always returns 100 as the value, but that doesn't seem to work, as the client mod continues to update the health correctly.

Here's the mixin:
```java
import org.spongepowered.asm.mixin.Mixin;
import org.spongepowered.asm.mixin.injection.At;
import org.spongepowered.asm.mixin.injection.Inject;
import org.spongepowered.asm.mixin.injection.callback.CallbackInfoReturnable;

import net.minecraft.network.packet.s2c.play.HealthUpdateS2CPacket;

@Mixin(HealthUpdateS2CPacket.class)
public abstract class HealthUpdateS2CPacketMixin {
    @Inject(at = @At("HEAD"), method = "Lnet/minecraft/network/packet/s2c/play/HealthUpdateS2CPacket;getHealth()F", cancellable = true)
    private float healthCancel(CallbackInfoReturnable<Float> cir) {
        return 100;
    }
}
```

**Comment by Daomephsta** [comment]
Mixin adds a *call* to your handler method (`healthCancel`) into the target method (`HealthUpdateS2CPacket.getHealth()`), not the body of your handler method. Thus a handler method returning does not cause the target method to return!  
Otherwise any handler method would force the target method to return, as all methods return, either explicitly or implicitly.

To force the target method to return, use the `setReturnValue` or `cancel` method of `CallbackInfoReturnable` and set `Inject.cancellable` to true.

Also you can't have `@Inject` handler methods that aren't void. So that mixin won't even allow the server to launch.

**Reply by TechPro424** [reply]
```java
@Inject(at = @At("HEAD"), method = "Lnet/minecraft/network/packet/s2c/play/HealthUpdateS2CPacket;getHealth()F", cancellable = true)
    private void healthCancel(CallbackInfoReturnable<Float> cir) {
        cir.setReturnValue(100.0f);
    }
```
I changed the inject to `void` return type and used `cir.setReturnValue`, still doesn't seem to make a difference

**Comment by KosmX** [comment]
You don't understand how these packet classes work:

**Server-side** [you need to patch here for anti-cheat]  
> The server logic constructs a packet using constructor (or setters) then convert to binary using `write(PacketByteBuf buf)`.  
> The server will send the binary data (with encription and optionally compression) to the client  

**Client-side** [you need to patch here for cheats, utility-client]
> Packet is constructed using `<init>(PacketByteBuf)`  
> `packet.apply` is invoked, it will read the values using getters. 

What you're trying to modify is the getter, what is used on client-side.  
Probably you want to inject into the constructor `HealthUpdateS2CPacket(float health, int food, float saturation)` and change the value on server-side.

**Reply by KosmX** [reply]
But be careful with some entities: tamed wolves [dogs] have a tail as a vanilla health indicator and you can break that if you fake dog's health.  
Also, for dead entities (0 hp) you should return the correct amount.

**Reply by TechPro424** [reply]
```java
@Mixin(HealthUpdateS2CPacket.class)
public abstract class HealthUpdateS2CPacketMixin {
    @Final @Mutable @Shadow private float health;
    @Final @Mutable @Shadow private int food;
    @Final @Mutable @Shadow private float saturation;


    @Inject(at = @At("HEAD"), method = "Lnet/minecraft/network/packet/s2c/play/HealthUpdateS2CPacket;<init>(FIF)V", cancellable = true)
    private void healthUpdateS2CPacket(float health, int food, float saturation, CallbackInfo ci) {
        this.health = 1000f;
        this.food = 1000;
        this.saturation = 1000f;
        ci.cancel();
    }
}
```

I injected this into the constructor, and the server started normally. But the moment I attempt to connect to the server, this error is thrown: https://gist.github.com/TechPro424/4729abc55cc4094af27820da31ad1075

**Reply by KosmX** [reply]
First learn basic mixin rules...  
You can only use TAIL or RETURN if you target a constructor. And you don't need to cancel anyway.

**Reply by TechPro424** [reply]
Changed the inject to target "TAIL" and removed `cancellable=true`

New inject:
```java
    @Inject(at = @At("TAIL"), method = "Lnet/minecraft/network/packet/s2c/play/HealthUpdateS2CPacket;<init>(FIF)V")
    private void healthUpdateS2CPacket(float health, int food, float saturation, CallbackInfo ci) {
        this.health = 1000f;
        this.food = 1000;
        this.saturation = 1000f;
    }
```

Now I'm able to enter the game, but client mod continues to update the health correctly.

**Reply by KosmX** [reply]
let's see: HealthUpdateS2CPacket is used for the player only...  
Entity health is stored using DataTracker   
```java
    private static final TrackedData<Float> HEALTH = DataTracker.registerData(LivingEntity.class, TrackedDataHandlerRegistry.FLOAT);
```
checking [wiki.vg](https://wiki.vg/) (MC protocol wiki) you can see that entity HP is synced using Set Entity Metadata packet with data ID `9`  
yarn name: `EntityTrackerUpdateS2CPacket`

You'll have to modify that on server-side to fake entity health.

**Reply by TechPro424** [reply]
I literally have no idea where to start on this one... probably just me being bad at this though
